ML project/Customer_churn_prediction.py

```diff
@@ -0,0 +1,512 @@
+
+
+# In[17]:
+
+
+df_scaled.churn.value_counts(normalize=True)
+
+
+# The target column shows that the dataset is unbalanced, hence stratifying based on this column would be a good choice
+
+# ### KNearestNeighbors
+
+# In[18]:
+
+
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import accuracy_score, roc_curve, roc_auc_score, f1_score, balanced_accuracy_score, precision_recall_curve
+from sklearn.model_selection import RandomizedSearchCV, cross_val_score
+
+
+# In[19]:
+
+
+X = df_scaled.drop('churn',axis=1).values
+y = df_scaled.churn.values
+
+
+# In[20]:
+
+
+X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,stratify=y)
+
+
+# In[21]:
+
+
+params = {'n_neighbors':[i for i in range(1,11)]}
+knc = KNeighborsClassifier()
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rcv = RandomizedSearchCV(knc, param_distributions = params, cv=3,scoring = i,random_state=123)
+    rcv.fit(X_train,y_train)
+    print('Metrics: {}\nBest params: {}\nBest score: {}'.format(i,rcv.best_params_,rcv.best_score_))
+    
+
+
+# In[22]:
+
+
+knc_f1 = []
+knc_b_acc = []
+knc_acc = []
+knc_auc = []
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    knc = KNeighborsClassifier(n_neighbors=10)
+    metrics = cross_val_score(knc, X_train, y_train, cv = 10, scoring=i)
+    if i == 'f1':
+        knc_f1.append(metrics)
+    elif i == 'balanced_accuracy':
+        knc_b_acc.append(metrics)
+    elif i == 'roc_auc':
+        knc_auc.append(metrics)
+    else:
+        knc_acc.append(metrics)
+
+
+# In[23]:
+
+
+plt.bar(['F1','Balanced accuracy','AUC','Accuracy'],[np.mean(knc_f1), np.mean(knc_b_acc), np.mean(knc_auc),np.mean(knc_acc)])
+plt.title('KNC Metric score (average)');
+
+
+# In[24]:
+
+
+knc_train_acc = []
+knc_test_acc = []
+
+for k in range (1,21):
+    knc = KNeighborsClassifier(n_neighbors=k)
+    knc.fit(X_train,y_train)
+    train_preds = knc.predict(X_train)
+    test_preds = knc.predict(X_test)
+    train_acc = accuracy_score(y_train,train_preds)
+    test_acc = accuracy_score(y_test,test_preds)
+    knc_train_acc.append(train_acc)
+    knc_test_acc.append(test_acc)
+
+
+# In[25]:
+
+
+plt.plot(range(1,21),knc_train_acc, '-rp')
+plt.plot(range(1,21),knc_test_acc, '-bp')
+plt.xlabel('N_neighbors')
+plt.ylabel('Accuracy')
+plt.xticks(range(1,21))
+plt.title('Train vs Test Accuracy with KNC',{'fontsize':15});
+
+
+# With K = 8 train accuracy and test accuracy are very close so the model does not seem to overfit
+
+# In[26]:
+
+
+knc = KNeighborsClassifier(n_neighbors=8)
+knc.fit(X_train,y_train)
+preds_k_8 = knc.predict(X_test)
+preds_prob_k_8 = knc.predict_proba(X_test)[:,1]
+fpr_k_8, tpr_k_8, tresholds_k_8 = roc_curve(y_test,preds_prob_k_8)
+prec_k_8, recall_k_8, tresh_k_8 = precision_recall_curve(y_test,preds_prob_k_8)
+auc_score_k_8 = roc_auc_score(y_test,preds_prob_k_8)
+
+
+# ### DecisionTreeClassifier
```

Comment:
@Ettore1991 code below is very similar to above. Any idea on how to reduce redundant code, thus making it easier to add new classifier?